tests/unit/test_waiters.py

```diff
@@ -392,6 +392,53 @@ def test_unspecified_errors_propagate_error_code(self):
         with self.assertRaisesRegexp(WaiterError, error_message):
             waiter.wait()
 
+    def _assert_failure_state_error_raised(self, acceptors, responses, expected_msg):
+        config = self.create_waiter_config(
+            acceptors=acceptors)
+        operation_method = mock.Mock()
+        waiter = Waiter('MyWaiter', config, operation_method)
+        self.client_responses_are(
+            *responses,
+            for_operation=operation_method
+        )
+        with self.assertRaisesRegexp(WaiterError, expected_msg):
+            waiter.wait()
+
+    def test_waiter_failure_state_error(self):
+        test_cases = [
+            ([{'state': 'failure', 'matcher': 'path',
+               'argument': 'Foo', 'expected': 'FAILURE'}],
+             [{'Foo': 'FAILURE'}],
+             'FAILURE'),
+            ([{'state': 'failure', 'matcher': 'pathAll',
+               'argument': 'Tables[].State', 'expected': 'FAILURE'}],
+             [{'Tables': [{"State": "FAILURE"}]}],
+             'FAILURE'),
+            ([{'state': 'failure', 'matcher': 'pathAny',
+               'argument': 'Tables[].State', 'expected': 'FAILURE'}],
+             [{'Tables': [{"State": "FAILURE"}]}],
+             'FAILURE'),
+            ([{'state': 'failure', 'matcher': 'status', 'expected': 404}],
+             [{'ResponseMetadata': {'HTTPStatusCode': 404}}],
+             '404'),
+            ([{'state': 'failure', 'matcher': 'error', 'expected': 'FailError'}],
+             [{'Error': {'Code': 'FailError', 'Message': 'foo'}}],
+             'FailError'),
+            ([{'state': 'retry', 'matcher': 'error', 'expected': 'RetryMe'}],
+             [{'Success': False}]*4,
+             'Max attempts exceeded'),
+            ([{'state': 'retry', 'matcher': 'error', 'expected': 'RetryMe'}],
```

Comment:
Can we keep this test consistent with the one we deleted. That way we're sure we're not losing any edge cases.